derror/multierror.go

```diff
@@ -0,0 +1,70 @@
+package derror
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// MultiError is an `error` type that represents a list of errors.
+//
+// Yes, another one.  What's wrong with all of the existing ones?  Looking at the ones that Emissary
+// already imports:
+//
+//  - `github.com/asaskevich/govalidator.Errors` : (1) Doesn't implement .Is().  (2) Output is all
+//    on one line and hard to read.
+//
+//  - `github.com/getkin/kin-openapi/openapi3.MultiError` : Output is all on one line and hard to
+//    read.
+//
+//  - `github.com/prometheus/client_golang/prometheus.MultiError` : (1) Doesn't implement .Is().
+//    (2) Output is pretty good, but doesn't handle child errors having newlines in them.
+//
+//  - `google.golang.org/appengine.MultiError` : (1) Doesn't implement .Is().  (2) Output is
+//    "helpfully" summarized to only print the first error and then say "(and X other errors)".
+//
+//  - `k8s.io/apimachinery/pkg/util/errors.Aggregate` : (1) Output is all on one line and hard to
+//    read.  (2) Isn't a simple []error type, and so is clunky to work with.
+//
+//  - `sigs.k8s.io/controller-tools/pkg/loader.ErrList` : (1) Doesn't implement .Is().  (2) Output
+//    is all on one line and hard to read.
+//
+// Note that (like k8s.io/apimachinery/pkg/util/errors.Aggregate) we don't implement .As(), because
+// we buy in to the same justification:
+//
+//   // Errors.As() is not supported, because the caller presumably cares about a
+//   // specific error of potentially multiple that match the given type.
+type MultiError []error
+
+func (errs MultiError) Error() string {
+	switch len(errs) {
+	case 0:
+		// This should not happen.
+		return "(0 errors; BUG: this should not be reported as an error)"
+	case 1:
+		return errs[0].Error()
+	default:
+		var buf strings.Builder
+		fmt.Fprintf(&buf, "%d errors:", len(errs))
+		for i, err := range errs {
+			prefix := fmt.Sprintf("\n %d. ", i+1)
+			for j, line := range strings.SplitAfter(err.Error(), "\n") {
+				if j == 1 {
+					prefix = strings.Repeat(" ", len(prefix)-1)
+				}
+				buf.WriteString(prefix)
+				buf.WriteString(line)
+			}
```

Comment:
It took me a minute to convince myself that this wasn't going to have spurious newlines all over the place. I _think_ I was initially reading the `if j == 1` clause as a random special case, not a change in state in the loop.

For me personally it might be more clear to just always set `prefix` to the space version at the end of the loop. I don't know if that's worth it for anyone else though.

Reply:
For whatever reason, when writing this I had it in my head that I really want to minimize allocations. But I'm not sure I had a good reason for that being in my head. I'll add a comment.